File: i3rd/UI/GuideMasterViewController.m

#import "GuideMasterViewController.h"
#import "GuideDetailViewController.h"

NSString *const GuideMasterViewControllerNextBackgroundImageKey = @"NextBackgroundImage";

@interface GuideMasterViewController ()

@property (strong, nonatomic) UIImageView *bodyView;

- (IBAction)moreViewControllerUnwound:(UIStoryboardSegue *)segue;

- (void)displayDetailViewController:(GuideDetailViewController *)detailViewController withSelectedIndexPath:(NSIndexPath *)indexPath;

@end

@implementation GuideMasterViewController

- (UIImageView *)bodyView
{
    if (_bodyView == nil) {
        if ([[UIDevice currentDevice] userInterfaceIdiom] == UIUserInterfaceIdiomPhone) {
            CGFloat statusBarHeight = [[UIApplication sharedApplication] statusBarFrame].size.height;
            CGFloat navigationBarHeight = self.navigationController.navigationBar.bounds.size.height;
            CGRect bodyViewFrame = UIEdgeInsetsInsetRect(self.navigationController.view.bounds, UIEdgeInsetsMake(statusBarHeight + navigationBarHeight, 0.0, 0.0, 0.0));
            _bodyView = [[UIImageView alloc] initWithFrame:bodyViewFrame];
            _bodyView.contentMode = UIViewContentModeScaleAspectFit;
            _bodyView.backgroundColor = [UIColor whiteColor];
            [self.navigationController.view insertSubview:_bodyView atIndex:0];
        } else {
            UINavigationController *detailNavigationController = self.splitViewController.viewControllers[1];
            CGFloat navigationBarHeight = detailNavigationController.navigationBar.bounds.size.height;
            CGRect bodyViewFrame = UIEdgeInsetsInsetRect(detailNavigationController.view.bounds, UIEdgeInsetsMake(navigationBarHeight, 0.0, 0.0, 0.0));
            _bodyView = [[UIImageView alloc] initWithFrame:bodyViewFrame];
            _bodyView.autoresizingMask = UIViewAutoresizingFlexibleBottomMargin | UIViewAutoresizingFlexibleHeight | UIViewAutoresizingFlexibleWidth;
            _bodyView.contentMode = UIViewContentModeScaleAspectFit;
            _bodyView.backgroundColor = [UIColor whiteColor];
            [detailNavigationController.view insertSubview:_bodyView atIndex:0];
        }
    }
    return _bodyView;
}

#pragma mark - Lifecycle

- (IBAction)moreViewControllerUnwound:(UIStoryboardSegue *)segue
{
    
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    self.navigationItem.title = @"攻略";

    NSURL *propertyListURL = [[[NSBundle mainBundle] bundleURL] URLByAppendingPathComponent:@"Guide.plist"];
    NSDictionary *propertyListInfo = [NSDictionary dictionaryWithContentsOfURL:propertyListURL];
    self.sections = [propertyListInfo objectForKey:PropertyListBasedViewControllerSectionsKey];
    
    if ([[UIDevice currentDevice] userInterfaceIdiom] == UIUserInterfaceIdiomPad) {
        NSIndexPath *indexPath = [NSIndexPath indexPathForRow:0 inSection:0];
        [self.tableView selectRowAtIndexPath:indexPath animated:NO scrollPosition:UITableViewScrollPositionNone];
        
        UINavigationController *detailNavigationController = self.splitViewController.viewControllers[1];
        GuideDetailViewController *detailViewController = (GuideDetailViewController *)detailNavigationController.topViewController;

        [self displayDetailViewController:detailViewController withSelectedIndexPath:indexPath];
    }
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
    if ([segue.identifier isEqualToString:@"ShowDetail"]) {
        NSIndexPath *indexPath = [self.tableView indexPathForSelectedRow];
        NSDictionary *sectionInfo = [self.sections objectAtIndex:indexPath.section];
        NSArray *rows = [sectionInfo objectForKey:PropertyListBasedViewControllerRowsKey];
        NSDictionary *rowInfo = [rows objectAtIndex:indexPath.row];
        NSString *rowTitle = [rowInfo objectForKey:PropertyListBasedViewControllerRowTitleKey];
        NSString *next = [rowInfo objectForKey:PropertyListBasedViewControllerNextKey];
        NSString *nextBackgroundImage = [rowInfo objectForKey:GuideMasterViewControllerNextBackgroundImageKey];
        
        NSURL *detailPropertyListURL = [[[NSBundle mainBundle] bundleURL] URLByAppendingPathComponent:next];
        NSDictionary *detailPropertyListInfo = [NSDictionary dictionaryWithContentsOfURL:detailPropertyListURL];
        NSArray *detailSections = [detailPropertyListInfo objectForKey:PropertyListBasedViewControllerSectionsKey];
        
        GuideDetailViewController *detailViewController = (GuideDetailViewController *)[segue.destinationViewController topViewController];
        detailViewController.title = rowTitle;
        detailViewController.sections = detailSections;
    }
}

- (void)displayDetailViewController:(GuideDetailViewController *)detailViewController withSelectedIndexPath:(NSIndexPath *)indexPath
{
    NSDictionary *sectionInfo = [self.sections objectAtIndex:indexPath.section];
    NSArray *rows = [sectionInfo objectForKey:PropertyListBasedViewControllerRowsKey];
    NSDictionary *rowInfo = [rows objectAtIndex:indexPath.row];
    NSString *rowTitle = [rowInfo objectForKey:PropertyListBasedViewControllerRowTitleKey];
    NSString *next = [rowInfo objectForKey:PropertyListBasedViewControllerNextKey];
    NSString *nextBackgroundImage = [rowInfo objectForKey:GuideMasterViewControllerNextBackgroundImageKey];
    
    NSURL *detailPropertyListURL = [[[NSBundle mainBundle] bundleURL] URLByAppendingPathComponent:next];
    NSDictionary *detailPropertyListInfo = [NSDictionary dictionaryWithContentsOfURL:detailPropertyListURL];
    NSArray *detailSections = [detailPropertyListInfo objectForKey:PropertyListBasedViewControllerSectionsKey];
    
    detailViewController.title = rowTitle;
    detailViewController.sections = detailSections;
    
//    self.bodyView.image = [UIImage imageNamed:nextBackgroundImage];
}

#pragma mark - State Restoration

#define GuideMasterViewControllerSelectedIndexPathKey @"SelectedIndexPath"

- (void)encodeRestorableStateWithCoder:(NSCoder *)coder
{
    [super encodeRestorableStateWithCoder:coder];
    
    [coder encodeObject:[self.tableView indexPathForSelectedRow] forKey:GuideMasterViewControllerSelectedIndexPathKey];
}

- (void)decodeRestorableStateWithCoder:(NSCoder *)coder
{
    [super decodeRestorableStateWithCoder:coder];
    
    NSIndexPath *selectedIndexPath = [coder decodeObjectForKey:GuideMasterViewControllerSelectedIndexPathKey];
    if (selectedIndexPath) {
        [self.tableView selectRowAtIndexPath:selectedIndexPath animated:NO scrollPosition:UITableViewScrollPositionTop];
        
        NSDictionary *sectionInfo = [self.sections objectAtIndex:selectedIndexPath.section];
        NSArray *rows = [sectionInfo objectForKey:PropertyListBasedViewControllerRowsKey];
        NSDictionary *rowInfo = [rows objectAtIndex:selectedIndexPath.row];
        NSString *nextBackgroundImage = [rowInfo objectForKey:GuideMasterViewControllerNextBackgroundImageKey];
        self.bodyView.image = [UIImage imageNamed:nextBackgroundImage];
    }
}

#pragma mark - Table View Delegate

- (void)tableView:(UITableView *)tableView willDisplayCell:(UITableViewCell *)cell forRowAtIndexPath:(NSIndexPath *)indexPath
{
    if ([[UIDevice currentDevice] userInterfaceIdiom] == UIUserInterfaceIdiomPad) {
        cell.accessoryType = UITableViewCellAccessoryNone;
    }
}

- (NSIndexPath *)tableView:(UITableView *)tableView willSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    if ([[UIDevice currentDevice] userInterfaceIdiom] == UIUserInterfaceIdiomPad && [[tableView cellForRowAtIndexPath:indexPath] isSelected]) {
        UINavigationController *detailNavigationController = self.splitViewController.viewControllers[1];
        [detailNavigationController popToRootViewControllerAnimated:YES];
        return nil;
    } else {
        return indexPath;
    }
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    if ([[UIDevice currentDevice] userInterfaceIdiom] == UIUserInterfaceIdiomPhone) {
        [self performSegueWithIdentifier:@"ShowDetail" sender:nil];
    } else {
        UINavigationController *detailNavigationController = self.splitViewController.viewControllers[1];
        [detailNavigationController popToRootViewControllerAnimated:NO];
        GuideDetailViewController *detailViewController = (GuideDetailViewController *)detailNavigationController.topViewController;
        [self displayDetailViewController:detailViewController withSelectedIndexPath:indexPath];
        detailViewController.tableView.contentOffset = CGPointZero;
        [detailViewController.tableView reloadData];
        [detailViewController.tableView flashScrollIndicators];
    }
}

@end

File: i3rd/UI/GuideMasterViewController.swift
//
//  GuideMasterViewController.swift
//  i3rd
//
//  Created by Leon Li on 2018/6/13.
//  Copyright © 2018 Leon & Vane. All rights reserved.
//

import UIKit

let GuideMasterViewControllerNextBackgroundImageKey = "NextBackgroundImage"

class GuideMasterViewController: PropertyListBasedViewController {
    lazy var bodyView: UIImageView = {
        if UIDevice.current.userInterfaceIdiom == .phone {
            let statusBarHeight = UIApplication.shared.statusBarFrame.height
            let navigationBarHeight = navigationController!.navigationBar.bounds.height
            let bodyViewFrame = UIEdgeInsetsInsetRect(navigationController!.view.bounds, UIEdgeInsets(top: statusBarHeight + navigationBarHeight, left: 0, bottom: 0, right: 0))
            let bodyView = UIImageView(frame: bodyViewFrame)
            bodyView.contentMode = .scaleAspectFit
            bodyView.backgroundColor = .white
            navigationController!.view.insertSubview(bodyView, at: 0)
            return bodyView
        } else {
            let detailNavigationController = splitViewController?.viewControllers[1] as! UINavigationController
            let navigationBarHeight = detailNavigationController.navigationBar.bounds.height
            let bodyViewFrame = UIEdgeInsetsInsetRect(navigationController!.view.bounds, UIEdgeInsets(top: navigationBarHeight, left: 0, bottom: 0, right: 0))
            let bodyView = UIImageView(frame: bodyViewFrame)
            bodyView.autoresizingMask = [.flexibleBottomMargin, .flexibleWidth, .flexibleHeight]
            bodyView.contentMode = .scaleAspectFit
            bodyView.backgroundColor = .white
            detailNavigationController.view.insertSubview(bodyView, at: 0)
            return bodyView
        }
    }()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        navigationItem.title = "攻略"
        
        let propertyListURL = Bundle.main.bundleURL.appendingPathComponent("Guide.plist")
        let propertyListInfo = NSDictionary(contentsOf: propertyListURL)
        sections = propertyListInfo?[PropertyListBasedViewControllerSectionsKey] as? [Any]
        
        if UIDevice.current.userInterfaceIdiom == .pad {
            let indexPath = IndexPath(row: 0, section: 0)
            tableView.selectRow(at: indexPath, animated: false, scrollPosition: .none)
            
            let detailNavigationController = splitViewController?.viewControllers[1] as! UINavigationController
            let detailViewController = detailNavigationController.topViewController as! GuideDetailViewController
            
            displayDetailViewController(detailViewController, withSelectedIndexPath: indexPath)
        }
    }
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        if segue.identifier == "ShowDetail" {
            let indexPath = tableView.indexPathForSelectedRow!
            let sectionInfo = sections[indexPath.section] as? NSDictionary
            let rows = sectionInfo?[PropertyListBasedViewControllerRowsKey] as? NSArray
            let rowInfo = rows?[indexPath.row] as? NSDictionary
            let rowTitle = rowInfo?[PropertyListBasedViewControllerRowTitleKey] as? String
            let next = rowInfo?[PropertyListBasedViewControllerNextKey] as? String
            
            let detailPropertyListURL = Bundle.main.bundleURL.appendingPathComponent(next!)
            let detailPropertyListInfo = NSDictionary(contentsOf: detailPropertyListURL)
            let detailSections = detailPropertyListInfo?[PropertyListBasedViewControllerSectionsKey] as? NSArray
            
            let detailViewController = (segue.destination as! UINavigationController).topViewController as! GuideDetailViewController
            detailViewController.title = rowTitle
            detailViewController.sections = detailSections as? [Any]
        }
    }
    
    @IBAction func moreViewControllerUnwound(_ segue: UIStoryboardSegue) {
        
    }
    
    func displayDetailViewController(_ detailViewController: GuideDetailViewController, withSelectedIndexPath indexPath: IndexPath) {
        let sectionInfo = sections[indexPath.section] as? NSDictionary
        let rows = sectionInfo?[PropertyListBasedViewControllerRowsKey] as? NSArray
        let rowInfo = rows?[indexPath.row] as? NSDictionary
        let rowTitle = rowInfo?[PropertyListBasedViewControllerRowTitleKey] as? String
        let next = rowInfo?[PropertyListBasedViewControllerNextKey] as? String
        let nextBackgroundImage = rowInfo?[GuideMasterViewControllerNextBackgroundImageKey] as? String
        
        let detailPropertyListURL = Bundle.main.bundleURL.appendingPathComponent(next!)
        let detailPropertyListInfo = NSDictionary(contentsOf: detailPropertyListURL)
        let detailSections = detailPropertyListInfo?[PropertyListBasedViewControllerSectionsKey] as? NSArray
        
        detailViewController.title = rowTitle
        detailViewController.sections = detailSections as? [Any]
        
        bodyView.image = UIImage(named: nextBackgroundImage!)
    }
    
    // MARK: - State Restoration
    
    let GuideMasterViewControllerSelectedIndexPathKey = "SelectedIndexPath"
    
    override func encodeRestorableState(with coder: NSCoder) {
        super.encodeRestorableState(with: coder)
        
        coder.encode(tableView.indexPathForSelectedRow, forKey: GuideMasterViewControllerSelectedIndexPathKey)
    }
    
    override func decodeRestorableState(with coder: NSCoder) {
        super.decodeRestorableState(with: coder)
        
        if let selectedIndexPath = coder.decodeObject(forKey: GuideMasterViewControllerSelectedIndexPathKey) as? IndexPath {
            tableView.selectRow(at: selectedIndexPath, animated: false, scrollPosition: .top)
            
            let sectionInfo = sections[selectedIndexPath.row] as? NSDictionary
            let rows = sectionInfo?[PropertyListBasedViewControllerRowsKey] as? NSArray
            let rowInfo = rows?[selectedIndexPath.row] as? NSDictionary
            let nextBackgroundImage = rowInfo?[GuideMasterViewControllerNextBackgroundImageKey] as? String
            bodyView.image = UIImage(named: nextBackgroundImage!)
        }
    }
    
    // MARK: - Table View Delegate
    
    override func tableView(_ tableView: UITableView, willDisplay cell: UITableViewCell, forRowAt indexPath: IndexPath) {
        if UIDevice.current.userInterfaceIdiom == .pad {
            cell.accessoryType = .none;
        }
    }
    
    override func tableView(_ tableView: UITableView, willSelectRowAt indexPath: IndexPath) -> IndexPath? {
        if UIDevice.current.userInterfaceIdiom == .pad && tableView.cellForRow(at: indexPath)?.isSelected == true {
            let detailNavigationController = splitViewController?.viewControllers[1] as! UINavigationController
            detailNavigationController.popToRootViewController(animated: true)
            return nil
        } else {
            return indexPath
        }
    }
    
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        if UIDevice.current.userInterfaceIdiom == .phone {
            performSegue(withIdentifier: "ShowDetail", sender: nil)
        } else {
            let detailNavigationController = splitViewController?.viewControllers[1] as! UINavigationController
            detailNavigationController.popToRootViewController(animated: false)
            let detailViewController = detailNavigationController.topViewController as! GuideDetailViewController
            displayDetailViewController(detailViewController, withSelectedIndexPath: indexPath)
            detailViewController.tableView.contentOffset = .zero
            detailViewController.tableView.reloadData()
            detailViewController.tableView.flashScrollIndicators()
        }
    }
}
